File: examples/05-fork-join/Example8.java
// =================================================================
//
// File: Example8.java
// Author(s): Sandra Tello A01703658 Isaac Planter A01702962
// Speedup: 1.10
// Description: This file implements the enumeration sort algorithm
// 				using Java's Fork-Join.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================
import java.util.concurrent.RecursiveAction;
import java.util.concurrent.ForkJoinPool;

public class Example8 extends RecursiveAction {
	private static final int SIZE = 100_000;
	private static final int MIN = 1_000;
	private int array[], start, end, aux[];


	public Example8(int array[], int start, int end, int aux[]) {
		this.array = array;
		this.start = start;
		this. end = end;
		this.aux=aux;
	}

	public void computeDirectly() {
		// place yout code here
		for (int i = start; i < end; i++) {
			aux[i] = 0;
		}
		for(int i=start;i<end;i++){
			for(int j=0;j<SIZE;j++){
				if(array[i]>array[j] | (array[i] == array[j] && j < i)){
					aux[i] += 1;
				}
			}
		}
		for (int k = start; k < end; k++) {
			array[aux[k]] = array[k];
		}
	}


	@Override
	protected void compute() {
		if ( (end - start) <= MIN ) {
			computeDirectly();
		} else {
			int mid = start + ((end - start) / 2);
			invokeAll(new Example8(array, start, mid, aux),
					  new Example8(array, mid, end, aux));
		}
	}

	public static void main(String args[]) {
		long startTime, stopTime;
		int array[] = new int[SIZE];
		int aux[]=new int[SIZE];
		double ms;
		ForkJoinPool pool;

		Utils.randomArray(array);
		Utils.displayArray("before", array);

		
		// int pos = Math.abs(Utils.r.nextInt()) % SIZE;
		// System.out.printf("Setting value 0 at %d\n", pos);
		// array[pos] = 0;

		System.out.printf("Starting with %d threads...\n", Utils.MAXTHREADS);
		ms = 0;
		for (int i = 0; i <= Utils.N; i++) {
			startTime = System.currentTimeMillis();

			pool = new ForkJoinPool(Utils.MAXTHREADS);
			pool.invoke(new Example8(array, 0, array.length, aux));

			stopTime = System.currentTimeMillis();
			ms += (stopTime - startTime);

		}

		Utils.displayArray("after", array);
		System.out.printf("avg time = %.5f\n", (ms / Utils.N));
	}

  // place your code here

}

File: examples/06-openmp/example8.c
// =================================================================
//
// File: example7.c
// Author(s): Sandra Tello A01703658 Isaac Planter A01702962
// Speedup: 4.07
// Description: This file contains the code that implements the
//				enumeration sort algorithm using OpenMP.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "utils.h"

#define SIZE 10000

void enumerationSort (int *arr, int size){

    int aux[size];
    #pragma omp parallel for shared(arr, size)
        for (int i = 0; i < size; i++){
            int menores = 0;
            for (int j = 0; j< size; j++){
                if(arr[i]>arr[j] || (arr[i]==arr[j]&&i<j) ){
                    menores+=1;
                }
            }
            aux[menores] = arr[i];
        }
        for (int k = 0; k<size; k++){
            arr[k] = aux[k];
        }

}

int main(int argc, char* argv[]) {
	int i, *a;
	double ms;

	a = (int*) malloc(sizeof(int) * SIZE);
	random_array(a, SIZE);
	display_array("before", a);

	printf("Starting...\n");
	ms = 0;
	for (i = 0; i < N; i++) {
		start_timer();

		// call the implemented function
		enumerationSort(a, SIZE);
		ms += stop_timer();
	}
	display_array("after", a);
	printf("avg time = %.5lf ms\n", (ms / N));

	free(a);
	return 0;
}
// implement your code
